WSD/src/cass/wsd/WSD.java

```diff
@@ -58,13 +66,101 @@ public WSD(String leftContext, String target, String rightContext, Language lang
 			// score is cardinality of intersection
 			int score = glossSet.size();
 			
-			scoredSynsets.add(new ScoredSense(sense, score));
+			scoredSenses.add(new ScoredSense(targetSense, score));
 		}
 		
 		// sort in descending order
-		Collections.sort(scoredSynsets);
-		Collections.reverse(scoredSynsets);
+		Collections.sort(scoredSenses);
+		Collections.reverse(scoredSenses);
+		
+		return scoredSenses;
+	}
+	
+	private List<ScoredSense> rankSensesUsingStochasticHypernymDistance() {
+		List<ScoredSense> scoredSenses= new ArrayList<ScoredSense>();
+		
+		for (WordSense targetSense : targetSenses) {
+			int senseScore = 0;
+			for (String contextWord : context) {
+				Set<WordSense> contextWordSenses = lTool.getSenses(contextWord);
+				int bestScore = 0;
+				for (WordSense contextWordSense : contextWordSenses) {
+					int currentScore = getHypernymDistanceScore(targetSense, contextWordSense);
+					if (currentScore > bestScore) {
+						bestScore = currentScore;
+					}
+				}
+				senseScore += bestScore;
+			}
+			scoredSenses.add(new ScoredSense(targetSense, senseScore));
+		}
+		
+		Collections.sort(scoredSenses);
+		Collections.reverse(scoredSenses);
+		
+		return scoredSenses;
+	}
+	
+	private int getHypernymDistanceScore(WordSense sense1, WordSense sense2) {
 		
-		return scoredSynsets;
+		List<WordSense> ancestors1 = getHypernymAncestors(sense1);
+		List<WordSense> ancestors2 = getHypernymAncestors(sense2);
+		
+		Collections.reverse(ancestors1);
+		Collections.reverse(ancestors2);
+		
+		// common ancestor has same ancestor path in both lists
+		int depthOfCommonAncestor = 0;
+		for (int i = 0; i < Math.min(ancestors1.size(), ancestors2.size()); i++) {
+			if (ancestors1.get(i) != ancestors2.get(i)) {
+				depthOfCommonAncestor = i-1;
```

Comment:
gimmee a break, giiiimmee a break, giiiiiimmeee a break in that for loop bro :musical_note: